!!! FEATURE: High level Neos site import `site:importAll`

bwaidelich opened the pull request:
Add commands `site:importall`, `site:exportall` and `site:pruneall` to import export and prune all sites of the given content repository.  This will export events files and sites into a package (Resources/Private/Content) or any other specified target directory.

Further changes:
- Remove contentStreamId and workspaceName from exported events as those are overwritten on import anyways
- Remove commands `cr:import`, `cr:export` and `cr:prune` because those did not take the site nodes into account.

## Todos

- [ ] More inline docs
- [x] `site:export` command
- [x] Add `sites.json` to export
- [x] Add "check for empty content stream" processor (for better error handling)
- [x] Extract workspace creation from events processor (only create if exists)
- [x] Remove `workspaceName`  and `contentStreamId` from `events.jsonl`
- [x] ~Re-add lowlevel import/export service (to make review easier)~ I re-added the tests and test-tools for now, I would like to skip the services and cli commands for now (until the rest is "safe")
- [x] Make `LegacyMigrationService` use high-level import
- [x] `\Neos\ContentRepository\LegacyNodeMigration\NodeDataToEventsProcessor` (now `\Neos\ContentRepository\LegacyNodeMigration\Processors\EventExportProcessor`) has to work with *workspaceName*  instead of *contentStreamId* -> [solved](https://github.com/neos/neos-development-collection/pull/5307#issuecomment-2436145225)

Resolves: #4448

bwaidelich committed:
WIP: FEATURE: Site import

Resolves: #4448

bwaidelich commented on Neos.ContentRepository.Export/src/Asset/AssetExporter.php:
```diff
@@ -14,7 +16,8 @@ public function __construct(
         private readonly ResourceLoaderInterface $resourceLoader,
-    ) {}
+    ) {
+    }
```
I sneaked in some cosmetic fixes to allow us to enable PHPCS for this package, too

bwaidelich commented on Neos.ContentRepository.Export/src/ExportService.php:
I decided to remove those low-level services for now.
It is quite easy to create them if needed (and more configurable then) but for Neos we'll need something more "highlevel" anyways

bwaidelich commented on Neos.ContentRepository.Export/src/ProcessingContext.php:
I know, "context" is not the best pattern, but in this case it really makes sense IMO:
It's a common object that is passed through all processors allowing them a shared access to the (virtual) filesystem and to dispatch "events"

(needs doc comments)

bwaidelich commented on Neos.ContentRepository.Export/src/Processors.php:
Just a plain (for now) collection object for processors that allows us to pass them around

bwaidelich commented on Neos.Neos/Classes/Domain/Service/SiteImportServiceFactory.php:
```diff
@@ -0,0 +1,67 @@
+            'Run doctrine migrations' => new DoctrineMigrateProcessor($this->doctrineService),
+            'Setup content repository' => new ContentRepositorySetupProcessor($serviceFactoryDependencies->contentRepository),
+            'Create Neos sites' => new SiteCreationProcessor($this->siteRepository),
+            // TODO create live workspace, etc
```
live workspace is currently created with the `EventStoreImportProcessor` but IMO the `RootWorkspaceWasCreated` event should be emitted in a separate processor

mficzel commented:
In #5306 we just moved cr:prune to the CrCommandController into the Neos.Neos->CrCommandController Package that gets deleted here.

bwaidelich committed:
Extract workspace creation to separate processor

bwaidelich committed:
Make `SiteImportService` a singleton

And don't rely on the `ContentRepositoryServiceFactoryInterface`

bwaidelich committed:
Merge branch '9.0' into feature/4448-site-import

bwaidelich committed:
Fix `LegacyMigrationService` initialization

bwaidelich committed:
Re-add `Neos.ContentRepository.Export` behat tests

and centralize common behat features into Trait

bwaidelich committed:
Rename `Factories` folder to `Factory

bwaidelich committed:
Remove unused namespace imports

mficzel committed:
Verify tha that Live workspace contains no events prior to importing

mficzel committed:
Ensure sites are persisted during import

mficzel committed:
Add site:export command again

mficzel committed:
Avoid exporting of contentStreamId

mficzel committed:
TASK: Make linter happy

mficzel committed:
Add replayall to site:import process

bwaidelich commented on Neos.Neos/Classes/Domain/Import/LiveWorkspaceIsEmptyProcessor.php:
```diff
@@ -0,0 +1,53 @@
+        $context->dispatch(Severity::NOTICE, 'Ensures empty live workspace');
+
+        if ($this->workspaceHasEvents(WorkspaceName::forLive())) {
+            throw new \RuntimeException( 'Live workspace already contains events please run "cr:prune" before importing.');
```
as discussed: This might become `site:removeAll`

mficzel replied:
adjusted

bwaidelich commented on Neos.Neos/Classes/Command/SiteCommandController.php:
```diff
@@ -131,11 +138,13 @@ public function createCommand($name, $packageKey, $nodeType, $nodeName = null, $
+     * Import sites
      *
-     * This command allows for importing one or more sites or partial content from the file system. The format must
+     * This command allows importing sites from the given path/packahe. The format must
```
typo :)

mficzel replied:
adjusted

bwaidelich commented on Neos.Neos/Classes/Command/SiteCommandController.php:
```diff
@@ -190,6 +199,58 @@ public function importCommand(string $packageKey = null, string $path = null, st
+     */
+    public function exportCommand(string $packageKey = null, string $path = null, string $contentRepository = 'default', bool $verbose = false): void
+    {
+        $exceedingArguments = $this->request->getExceedingArguments();
+        if (isset($exceedingArguments[0]) && $packageKey === null && $path === null) {
+            if (file_exists($exceedingArguments[0])) {
+                $path = $exceedingArguments[0];
+            } elseif ($this->packageManager->isPackageAvailable($exceedingArguments[0])) {
+                $packageKey = $exceedingArguments[0];
+            }
+        }
+        if ($packageKey === null && $path === null) {
+            $this->outputLine('<error>You have to specify either <em>--package-key</em> or <em>--filename</em></error>');
+            $this->quit(1);
+        }
```
nitpick: this is the same code as for `importCommand`, could be extracted to a private method

mficzel replied:
adjusted

bwaidelich commented on Neos.Neos/Classes/Domain/Service/SiteImportService.php:
```diff
@@ -0,0 +1,88 @@
+            'Create Live workspace' => new LiveWorkspaceCreationProcessor($contentRepository, $this->workspaceService),
+            'Import events' => $this->contentRepositoryRegistry->buildService($contentRepositoryId, new EventStoreImportProcessorFactory(WorkspaceName::forLive(), keepEventIds: true)),
+            'Import assets' => new AssetRepositoryImportProcessor($this->assetRepository, $this->resourceRepository, $this->resourceManager, $this->persistenceManager),
+            'Replay all projections' => $this->contentRepositoryRegistry->buildService($contentRepositoryId, $this->projectionReplayServiceFactory),
```
This should not be required.
If it _is_, we'll have to investigate. But the plan is to publish events just like the CR would do.
If the catchups are skipped, we could invoke _those_. But a full replay would be an overkill

mficzel replied:
switched to catchup only. Replayall is still needed after prune i think.

bwaidelich commented on Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:
```diff
@@ -24,10 +24,13 @@ public function __construct(
     {
+        $payload = \json_decode($event->data->value, true, 512, JSON_THROW_ON_ERROR);
+        // unset content stream id as this is overwritten during import
+        unset($payload['contentStreamId']);
```
IMO we can/should remove the workspace name, too:
```suggestion
        unset($payload['contentStreamId'], $payload['workspaceName']);
```

mficzel replied:
adjusted

bwaidelich reviewed (commented):
Looking great already, thanks for taking over. Just a few remarks

mficzel committed:
Rename command `site:import` and `site:export` to `site:importAll` and `site:exportAll` to be in line with current behavior

mficzel committed:
Add `site:pruneAll` command that empties the current cr and removes all referenced site records

mficzel committed:
Improve after review

- centralize pathDetermination and closure generationIn site command controller
- use catch-up instead of replayall after import
- remove workspace name from exported events

mficzel committed:
Extract SitePruningService

mficzel commented:
Wonder about "Make LegacyMigrationService use high-level import" ... this would make the legacy migration depend on Neos.Neos ... technically there is no non Neos way to get an old Neos CR but it seems not that elegant anyways.

mficzel committed:
Remove ProcessorInterface from ProjectionService and introduce separate ProjectionCatchUpProcessor and ProjectionReplayProcessor.

mficzel committed:
Refactor migration of legacyData and add additional command `cr:exportLegacyData` alongside `cr:migrateLegacyData`

bwaidelich commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Command/CrCommandController.php:
```diff
@@ -38,16 +38,16 @@
+        private readonly PropertyMapper              $propertyMapper,
+        private readonly ContentRepositoryRegistry   $contentRepositoryRegistry,
+        private readonly SiteRepository              $siteRepository,
+        private readonly ProjectionServiceFactory    $projectionServiceFactory,
```
Not a big issue ofc, but I don't like the style of aligning parameters (mostly because it leads to multi-line diffs like this one) and we don't do that in other places

mficzel replied:
I am not sure this is an issue at all. The export does not contain neither workspaceName nor contentStreamId any more so those values are only set on the exported events as temporary values so we can use the default eventObjects.

mficzel replied:
Not sure what the previous comment was for. I also do not like those aligned parameters ... sometimes phpstorm goes rogue and adds them without my consent. I removed that now, also the controller is now the SiteController

bwaidelich reviewed (commented):
@mficzel This is great, thanks!
Just a nitpicky comment and the last todo is still open, right?
> `\Neos\ContentRepository\LegacyNodeMigration\NodeDataToEventsProcessor` has to work with workspaceName instead of contentStreamId

mficzel commented:
I also think we should rename the CrCommandController in the Legacymigration Package to SiteCommandController so the commands become `site:migrateLegacyData` and `site:exportLegacyData`. This is justified as we are clearly handling more than just the cr with the Sites and domains.

mficzel committed:
Rename  `cr:migratelegacydata`, `cr:exportlegacydata`, to `site:migratelegacydata`, `site:exportlegacydata` and adjust the readmes

mficzel commented:
I will check the behat tests for the import export stuff now. Should be the last thing to adjust.

mficzel committed:
Adjust behat tests for legacy export and add sites.feature

mficzel commented:
Converted to draft for now because of a merge conflict and probably some issues in the behat test that still need adjustment.

Could use some help there

mficzel committed:
Format arguments of site command controller properly

mficzel commented:
@bwaidelich 

>  the last todo is still open, right?
>>  \Neos\ContentRepository\LegacyNodeMigration\NodeDataToEventsProcessor has to work with workspaceName instead of contentStreamId

I am not sure this is an issue at all. The export does not contain neither workspaceName nor contentStreamId any more so those values are only set on the exported events as temporary values so we can use the default eventObjects.

bwaidelich commented:
>  The export does not contain neither workspaceName nor contentStreamId any more 

Ah of course, you're right!

mhsdesign commented on Neos.ContentRepository.Export/src/Event/ValueObject/ExportedEvent.php:
```diff
@@ -24,10 +24,13 @@ public function __construct(
     public static function fromRawEvent(Event $event): self
     {
+        $payload = \json_decode($event->data->value, true, 512, JSON_THROW_ON_ERROR);
+        // unset content stream id as this is overwritten during import
+        unset($payload['contentStreamId'], $payload['workspaceName']);
```
at some point i think we should introduce a utility to manifest this stuff :D we do it at a lot of places to assume the properties are named that way ^^

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Processors/SitesExportProcessor.php:
```diff
@@ -0,0 +1,68 @@
+use Neos\ContentRepository\Export\ProcessorInterface;
+
+/**
+ * @phpstan-type DomainShape array{hostname: string, scheme?: ?string, port?: ?int, active?: ?bool, primary?: ?bool }
+ * @phpstan-type SiteShape array{name:string, siteResourcesPackageKey:string, nodeName?: string, online?:bool, domains?: ?DomainShape[] }
```
❤️

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/Service/ProjectionService.php:
```diff
@@ -14,13 +14,12 @@
+ * @internal this is currently only used by the {@see SiteCommandController}
  */
-final class ProjectionReplayService implements ContentRepositoryServiceInterface
+final class ProjectionService implements ContentRepositoryServiceInterface
```
fyi in a followup we might move this class into the cr core to catchup projections etc but its fine if it stays here for now

mhsdesign commented on Neos.Neos/Classes/Command/SiteCommandController.php:
```diff
@@ -220,4 +327,47 @@ function (Site $site) use ($siteNodePattern) {
+            $this->outputLine(match ($severity) {
+                Severity::NOTICE => $message,
+                Severity::WARNING => sprintf('<error>Warning: %s</error>', $message),
+                Severity::ERROR => sprintf('<error>Error: %s</error>', $message),
```
they are the same color :D

mhsdesign commented on Neos.Neos/Classes/Domain/Import/LiveWorkspaceIsEmptyProcessor.php:
```diff
@@ -0,0 +1,54 @@
+
+    private function workspaceHasEvents(WorkspaceName $workspaceName): bool
+    {
+        /** @phpstan-ignore-next-line internal method of the cr is called */
```
we should probably add a error identifier like `neos.internal` here so that we dont ignore other errors in the line

mhsdesign replied:
Thanks for the adjustments @mficzel ❤️ :)

mhsdesign commented on Neos.Neos/Classes/Domain/Import/LiveWorkspaceIsEmptyProcessorFactory.php:
```diff
@@ -0,0 +1,31 @@
+/**
+ * @implements ContentRepositoryServiceFactoryInterface<LiveWorkspaceIsEmptyProcessor>
+ */
+final readonly class LiveWorkspaceIsEmptyProcessorFactory implements ContentRepositoryServiceFactoryInterface
```
except this we can avoid accessing this internal things ... probably good enough? But still against https://github.com/neos/neos-development-collection/pull/5286 :D

mhsdesign replied:
maybe the logic could be part of the content repository or the `ContentStreamPruner` (eg rename that thing :D)

mhsdesign commented on Neos.Neos/Classes/Domain/Import/SiteCreationProcessor.php:
```diff
@@ -0,0 +1,123 @@
+/**
+ * Import processor that creates and persists a Neos {@see Site} instance
+ *
+ * @phpstan-type DomainShape array{hostname: string, scheme?: ?string, port?: ?int, active?: ?bool, primary?: ?bool }
+ * @phpstan-type SiteShape array{name:string, siteResourcesPackageKey:string, nodeName?: string, online?:bool, domains?: ?DomainShape[] }
```
one can import shapes via `@phpstan-import-type Value from SerializedPropertyValue`

mficzel replied:
Adjusted

mhsdesign commented on Neos.Neos/Classes/Domain/Pruning/ContentRepositoryPruningProcessor.php:
```diff
@@ -0,0 +1,49 @@
+    ) {
+    }
+
+    public function run(ProcessingContext $context): void
```
this is a little in conflict with whats being done in: https://github.com/neos/neos-development-collection/pull/5297

eg id say we should still use the `ContentStreamPruner` here so we can also avoid this hardcore logic in neos

mficzel replied:
Can we adjust that after #5297 is merged?

mhsdesign replied:
Jup after the merge ill propose a change for this branch `feature/4448-site-import` how i would wire it together, but nothing at the logic will change.

mhsdesign replied:
adjusted to use the `contentStreamPruner->pruneAllWorkspacesAndContentStreamsFromEventStream` here :) and no need to use the cr internals here as well ... as the cr needs to provide tools for this itself

mhsdesign commented on Neos.Neos/Classes/Domain/Pruning/RoleAndMetadataPruningProcessorFactory.php:
```diff
@@ -0,0 +1,39 @@
+use Neos\ContentRepository\Core\Factory\ContentRepositoryServiceInterface;
+use Neos\Neos\Domain\Service\WorkspaceService;
+
+/**
+ * @implements ContentRepositoryServiceFactoryInterface<RoleAndMetadataPruningProcessor>
+ */
+final readonly class RoleAndMetadataPruningProcessorFactory implements ContentRepositoryServiceFactoryInterface
```
i think this should not be a low level CR service

mficzel replied:
Where should this be handled then. Also this is in the Neos.Neos Package for exactly that reason as it is needed but not part of the cr itself.

mhsdesign replied:
i replaced it with a simple `new RoleAndMetadataPruningProcessor` ... we dont need the `ContentRepositoryServiceFactoryInterface` here for instantiation

mhsdesign commented on Neos.Neos/Classes/Domain/Import/DoctrineMigrateProcessor.php:
```diff
@@ -0,0 +1,32 @@
+
+    public function run(ProcessingContext $context): void
+    {
+        $this->doctrineService->executeMigrations();
```
doing that in the same process is unsafe when not closing and reopening the connections as a lot of transactions will be nested: see https://github.com/neos/behat/pull/40

mficzel replied:
This is for convenience mostly. Most people will have the migrations applied and it may help some.  I do not see where this would do more harm than good. If you try to import with incomplete migrations failure is guaranteed.

mhsdesign replied:
Ah i think you misunderstood me: My thesis this will not work as is but needs a $this->connection->close() hotfix to reset the transactions as otherwise the event store doesnt work (see https://github.com/neos/neos-development-collection/pull/5284) ...

mhsdesign replied:
and now that i think of it, maybe the user doesnt gain anything here as there is not that good output / error handling than if run via `doctrine:migrate`

and now there is a huge overlap to what `flow setup` advises to do and what is done automatically - if it works....

ill think about what `flow setup`'s responsibilities are and what is part of the site setup. https://github.com/neos/neos-setup

But i think setting up migrations is not responsibility (but we can surely check that beforehand and give helpful errors with run `flow setup` first or sth ...)

mficzel replied:
So you suggest to remove the doctrine migrations but keep cr:setup here. To me that would be fine.

mhsdesign replied:
As written in https://github.com/neos/neos-development-collection/pull/5307#issuecomment-2453048011 actually now that i think of it i would keep the fingers from adjusting the schema hear, and only warn if not done.

mhsdesign commented on Neos.Neos/Classes/Domain/Service/SiteImportService.php:
```diff
@@ -0,0 +1,92 @@
+        $context = new ProcessingContext($filesystem, $onMessage);
+        $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+
+        // TODO make configurable (?)
```
rather idea than todo ^^ :D

mficzel replied:
Agree this can be removed. I also think that we do not need extensibility here as anyone can create its own processing chain. If he is able to create a custom processor he can aswell do that.

mhsdesign reviewed (commented):
I had a quick look :) thanks for all the work and the new processing context and stuff looks good. Im not super happy with the heavy use of creating own cr services in Neos.Neos and looking into the event stream and deleting streams.

That is all responsibility of the cores `ContentStreamPruner` imo (which could also be renamed so its clear it also handles workspaces) and that thing will also work when im done with it *g https://github.com/neos/neos-development-collection/pull/5297

So it would need a method like `workspaceHasEvents` to cover what `LiveWorkspaceIsEmptyProcessor` needs and prune content AND workspace streams for `ContentRepositoryPruningProcessor`

mficzel committed:
Adjust after QS Feedback:

- Add identifier "neos.cr.internal" for ignored internal calls
- Reuse shapes from Export in Import
- Remove comment about extensibility via configuration

mhsdesign commented on Neos.Neos/Classes/Domain/Import/SiteCreationProcessor.php:
```diff
@@ -0,0 +1,124 @@
+                    'siteResourcesPackageKey' => self::extractPackageKeyFromNodeTypeName($event->payload['nodeTypeName']),
+                    'name' => $event->payload['initialPropertyValues']['title']['value'] ?? $event->payload['nodeTypeName'],
+                    'nodeTypeName' => $event->payload['nodeTypeName'],
+                    'nodeName' => $event->payload['nodeName'] ?? null,
```
`nodeName` must not be empty here.

mhsdesign commented on Neos.Neos/Classes/Domain/Import/SiteCreationProcessor.php:
```diff
@@ -0,0 +1,124 @@
+    /**
+     * @return array<SiteShape>
+     */
+    private static function extractSitesFromEventStream(ProcessingContext $context): array
```
uiuiuiuiuiu but i see this is rather only a fallback in case the `sites.json` is missing .. good idea :) i would still like to mark this as `@deprecated` and also indicate on the console that were attempting a fallback strategy here:

```
$context->dispatch(Severity::WARNING, "No sites.json in export found. Attempting to decode sites from events.");
```

and then also log what sites will be created?

mhsdesign commented on Neos.Neos/Classes/Domain/Service/SiteImportService.php:
```diff
@@ -0,0 +1,91 @@
+        $processors = [
+            'Run doctrine migrations' => new DoctrineMigrateProcessor($this->doctrineService),
+            'Setup content repository' => $this->contentRepositoryRegistry->buildService($contentRepositoryId, new ContentRepositorySetupProcessorFactory()),
+            'Verify Live workspace does not exist yet' => $this->contentRepositoryRegistry->buildService($contentRepositoryId, new LiveWorkspaceIsEmptyProcessorFactory()),
+            'Create Neos sites' => new SiteCreationProcessor($this->siteRepository, $this->domainRepository, $this->persistenceManager),
+            'Create Live workspace' => new LiveWorkspaceCreationProcessor($contentRepository, $this->workspaceService),
```
can we not just swap Create Neos site and Create Live workspace? That way the live workspace step could throw an exception instead of Verify Live workspace does not exist yet?

mhsdesign commented on Neos.Neos/Classes/Domain/Import/LiveWorkspaceIsEmptyProcessor.php:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    private function workspaceHasEvents(WorkspaceName $workspaceName): bool
```
This `workspaceHasEvents` might be a little to hardcore and we could instead just use `findWorkspaceByName` ... and if the workspace indeed exists in the events the creation should fail in `handleCreateRootWorkspace` (which it not does yet but should :D)

mhsdesign commented on Neos.Neos/Classes/Command/SiteCommandController.php:
```diff
@@ -111,31 +156,93 @@ public function createCommand($name, $packageKey, $nodeType, $nodeName = null, $
-            $this->outputLine('The site:prune command expects the "Node name" from the site list as a parameter.');
-            $this->outputLine('If you want to delete all sites, you can run <b>site:prune \'*\'</b>.');
-            $this->quit(1);
+        // Since this command uses a lot of memory when large sites are imported, we warn the user to watch for
+        // the confirmation of a successful import.
+        $this->outputLine('<b>This command can use a lot of memory when importing sites with many resources.</b>');
+        $this->outputLine('If the import is successful, you will see a message saying "Import of site ... finished".');
+        $this->outputLine('If you do not see this message, the import failed, most likely due to insufficient memory.');
+        $this->outputLine('Increase the <b>memory_limit</b> configuration parameter of your php CLI to attempt to fix this.');
+        $this->outputLine('Starting import...');
+        $this->outputLine('---');
```
ha this is copied from Neos 8.3 :D but im not sure if this is still true? In 8.3 the xml parsing was super expensive but now we just read json line by line and see ...

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/4448-site-import

mhsdesign committed:
TASK: Post merge adjustments to make phpstan happy

mhsdesign committed:
TASK: Replace `LiveWorkspaceIsEmptyProcessor` with read model checks

mhsdesign committed:
TASK: Make `RoleAndMetadataPruningProcessor` a simple processor

we dont want to use the internal cr service factory pattern here

mhsdesign committed:
TASK: Make `ContentRepositoryPruningProcessor` use the `ContentStreamPruner`

mhsdesign committed:
TASK: Turn `SitePruningProcessor` into simple processor without factory

mhsdesign committed:
TASK: Turn `SiteExportProcessor` into simple processor without factory

mhsdesign committed:
TASK: Turn `ProjectionCatchupProcessor` and `ProjectionReplayProcessor` into simple processor without factory

mhsdesign commented:
Okay i fixed the merge conflicts got rid of the low level event store access in Neos.Neos and now only the `EventExportProcessor` and `EventStoreImportProcessor` are - as required - low level cr services (`ContentRepositoryServiceInterface`) but nothing else :D

Additionally i think its not the purpose of the import to setup the schemas i think as this must better be handled in a separate command (timeout and also transaction nesting etc)

Thats why i would propose to remove `ContentRepositorySetupProcessor` and `DoctrineMigrateProcessor` and just check if they have been run beforehand. (see https://github.com/neos/neos-development-collection/pull/5307#discussion_r1819745466)

we cannot do everything in one ... where does it stop? setting up image configuration and database as well? :D

Thats why we have `flow setup` at all ... i would just throw an error if any of the preconditions are not met (no database, no schema ...) and tell people to use `flow setup` ... that is imo the entry point which will lead people to use the `site:import` _afterwards_

mhsdesign committed:
TASK: Document and improve `extractSitesFromEventStream`

mhsdesign committed:
TASK: Move todo around :)

mhsdesign committed:
TASK: Remove redundant documentation

mhsdesign committed:
TASK: Reintroduce resetting of projection state via `ProjectionResetProcessor`

as in `cr:prune`

mhsdesign committed:
TASK: Use `Processors::fromArray`

mhsdesign committed:
TASK: Fail early if `resource://` paths are used in export

see https://github.com/neos/neos-development-collection/issues/4912

mhsdesign commented on Neos.Neos/Classes/Command/SiteCommandController.php:
```diff
@@ -220,4 +327,51 @@ function (Site $site) use ($siteNodePattern) {
+            $package = $this->packageManager->getPackage($packageKey);
+            $path = Files::concatenatePaths([$package->getPackagePath(), 'Resources/Private/Content']);
+        }
+        if (str_starts_with($path, 'resource://')) {
+            $this->outputLine('<error>Resource paths are not allowed, please use <em>--package-key</em> instead or a real path.</error>');
+            $this->quit(1);
+        }
```
todo are we okay with ignoring this for now, or should we transform the `resource` path to a real path? https://github.com/neos/neos-development-collection/issues/4912

(https://github.com/neos/flow-development-collection/issues/2687)

mhsdesign commented on Neos.Neos/Classes/Domain/Import/LiveWorkspaceCreationProcessor.php:
```diff
@@ -0,0 +1,55 @@
+        $context->dispatch(Severity::NOTICE, 'Creating live workspace');
+        $liveWorkspace = $this->contentRepository->findWorkspaceByName(WorkspaceName::forLive());
+        $liveContentStreamVersion = $liveWorkspace ? $this->contentRepository->findContentStreamById($liveWorkspace->currentContentStreamId)?->version : null;
+        if ($liveWorkspace && $liveContentStreamVersion !== 0) {
+            // todo we cannot use `hasPublishableChanges` here... maybe introduce `hasChanges`?
+            throw new \RuntimeException('Live workspace already contains content please run "site:pruneAll" before importing.');
+        }
```
todo this is not totally ideal yet. Should we maybe adjust https://github.com/neos/neos-development-collection/pull/5332 to be just `hasChanges`?

mhsdesign replied:
I think this is no use-case we need to support? IF the workspace already exists we need to abort?

mhsdesign commented on Neos.ContentRepository.Export/Tests/Behavior/Features/Bootstrap/CrImportExportTrait.php:
```diff
@@ -167,6 +144,66 @@ public function iExpectTheFollowingJsonL(PyStringNode $string): void
+    /**
+     * @Then I expect the following sites to be exported
+     */
+    public function iExpectTheFollowingSitesToBeExported(TableNode $table): void
```
we should also test this `I expect the following sites to be exported` for Neos.Neos (not only the legacy migrations probably) ... and move this method to Neos.Neos then?

mhsdesign commented on Neos.ContentRepository.Export/src/Processors/AssetExportProcessor.php:
```diff
@@ -25,102 +26,71 @@
  */
 final class AssetExportProcessor implements ProcessorInterface
```
doesnt have to change here but why is this asset export part of he standalone `Export` package? :D

I mean it says so itself:

> Note: This processor requires the packages "neos/media" and "neos/neos" to be installed!

mhsdesign replied:
i think because it was all previously hard wired in the export package ... but now it could be moved to neos.neos

mhsdesign replied:
@bwaidelich what do you say were should `\Neos\ContentRepository\Export\Processors\AssetExportProcessor` now reside, it doesnt have to be in `Export` anymore as this now can be fully standalone and independent of neos.

bwaidelich replied:
theoretically it could be put in a completely separate package, but I guess we agree that that is an overkill :)
Can it be in `Neos.Media`? Otherwise, why not just keep it here? It does not create a dependency to `neos/media` it just says: _if_ you want to use this processor, you'll need `neos/media`

mhsdesign replied:
okay either way it will be a followup pr:) And it DOES imo and php architecture linting would fail

kitsunet replied:
cannot be in neos/media but could be in neos/neos. I still think it _should_ be there, this "if you want to use this you got to install XYZ" seems fishy

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Command/SiteCommandController.php:
```diff
@@ -83,71 +72,84 @@ public function migrateLegacyDataCommand(bool $verbose = false, string $config =
         $this->verifyDatabaseConnection($connection);
 
+        $contentRepositoryId = ContentRepositoryId::fromString($contentRepository);
+        $temporaryFilePath = $this->environment->getPathToTemporaryDirectory() . uniqid('Export', true);
+        Files::createDirectoryRecursively($temporaryFilePath);
```
instead of operating on the real file system we could also run this in memory but that would probably overflow? (with the in memory adapter)

kitsunet replied:
php://temp can overflow into a temporary file, that could be an option

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Command/SiteCommandController.php:
```diff
@@ -83,71 +72,84 @@ public function migrateLegacyDataCommand(bool $verbose = false, string $config =
+        );
 
-        $site = $this->siteRepository->findOneByNodeName($siteNodeName);
-        if ($site !== null) {
```
cool to have that cleaned up as well ❤️ was not even aware of that mess really ^^

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Command/SiteCommandController.php:
```diff
@@ -83,71 +72,84 @@ public function migrateLegacyDataCommand(bool $verbose = false, string $config =
+     * @param string|null $config JSON encoded configuration, for example '{"dbal": {"dbname": "some-other-db"}, "resourcesPath": "/some/absolute/path"}'
+     * @throws \Exception
+     */
+    public function exportLegacyDataCommand(string $path, bool $verbose = false, string $config = null): void
```
thats really a good thing to have!!!

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Command/SiteCommandController.php:
```diff
@@ -83,71 +72,84 @@ public function migrateLegacyDataCommand(bool $verbose = false, string $config =
+            $this->createOnProcessorClosure(),
+            $this->createOnMessageClosure($verbose)
+        );
 
```
there should be a message like: legacy export finished, attempting to import from temporary $temporaryFilePath

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Processors/EventExportProcessor.php:
```diff
@@ -132,14 +102,10 @@ public function run(): ProcessorResult
-            } catch (MigrationException $exception) {
-                return ProcessorResult::error($exception->getMessage());
+                $this->processNodeData($context, $nodeDataRow);
+            } catch (MigrationException $e) {
+                throw new \RuntimeException($e->getMessage(), 1729506899, $e);
             }
```
that makes no sense here and probably in other places? The `MigrationException` might as well just bubble up.

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Processors/EventExportProcessor.php:
```diff
@@ -369,7 +316,7 @@ public function extractPropertyValuesAndReferences(array $nodeDataRow, NodeType
 
             if (!$nodeType->hasProperty($propertyName)) {
-                $this->dispatch(Severity::WARNING, 'Skipped node data processing for the property "%s". The property name is not part of the NodeType schema for the NodeType "%s". (Node: %s)', $propertyName, $nodeType->name->value, $nodeDataRow['identifier']);
+                $context->dispatch(Severity::WARNING, "Skipped node data processing for the property \"{$propertyName}\". The property name is not part of the NodeType schema for the NodeType \"{$nodeType->name->value}\". (Node: {$nodeDataRow['identifier']})");
```
i like `sprintf` better ... but ... :D Please dont change it everywhere in Neos 😂

mhsdesign commented on Neos.Neos/Classes/Command/SiteCommandController.php:
```diff
@@ -111,31 +156,93 @@ public function createCommand($name, $packageKey, $nodeType, $nodeName = null, $
+     * be identical to that produced by the exportAll command.
      *
-     * @param string $siteNode Name for site root nodes to clear only content of this sites (globbing is supported)
+     * !!! The live workspace has to be empty prior to importing. !!!
```
```suggestion
     * The live workspace has to be empty prior to importing.
```

a bit too loud i cannot hear anymore.

mhsdesign reviewed (commented):
Thank you both for all the work. I read through most parts. The neos ui e2e tests will need adjustments but ill take care of that later.

mhsdesign committed:
TASK: Use `Processors::fromArray`

bwaidelich commented:
> Additionally i think its not the purpose of the import to setup the schemas i think as this must better be handled in a separate command (timeout and also transaction nesting etc)

I have a different tendency in this specific matter:
There's a set of _specific_ preconditions that have to be met in order to import a site and I think it is just a better UX if the import _detects_ whether any of those conditions are not met. And it's even a better UX if the import fixes this if it can.
IMO that's in no contradiction to lower-level commands like `cr:setup` and `doctrine:migrate` (that are expected to be invoked by machines usually, i.e. deployment script)

> we cannot do everything in one ... where does it stop? setting up image configuration and database as well? :D

In my head there is a clear distinction. All those "preparing" processors have to...:
- ...be idempotent 
- ...be relatively "cheap" if they lead to a no-op
- ...only affect data that is stored in the database (i.e. not change any settings for example)


Of course, having said that, it should always be weighted against the _costs_.
In the case of the [ContentRepositorySetupProcessor](https://github.com/neos/neos-development-collection/blob/c98384d434013e14d8ba56dab32e15b82437668b/Neos.ContentRepository.Export/src/Processors/ContentRepositorySetupProcessor.php) and the [DoctrineMigrateProcessor](https://github.com/neos/neos-development-collection/blob/c98384d434013e14d8ba56dab32e15b82437668b/Neos.Neos/Classes/Domain/Import/DoctrineMigrateProcessor.php) the cost is pretty low in terms of complexity.
However, as far as I see it, it's an issue to change the db schema and -data in the same db transaction. So the implicative cost is indeed higher..
I'd say: If we can easily solve the issue with a cloned db connection we should do that.

OTOH I am not mad if you decide to kill those two processors, obviously

bwaidelich commented on Neos.ContentRepository.Export/src/Processors/EventStoreImportProcessor.php:
```diff
@@ -106,74 +92,17 @@ public function run(): ProcessorResult
+        $contentStreamStreamName = ContentStreamEventStreamName::fromContentStreamId($workspace->currentContentStreamId)->getEventStreamName();
         try {
-            $contentStreamCreationCommitResult = $this->eventStore->commit($contentStreamStreamName, $events, ExpectedVersion::NO_STREAM());
+            $this->eventStore->commit($contentStreamStreamName, Events::fromArray($domainEvents), ExpectedVersion::ANY());
```
`ExpectedVersion::ANY()` is not desired here, we should change that to
```suggestion
            $this->eventStore->commit($contentStreamStreamName, Events::fromArray($domainEvents), ExpectedVersion::fromVersion(Version::first()));
```

mhsdesign committed:
TASK: Remove automatic schema creation from import and redirect user to `./flow cr:setup`

`./flow doctrine:migrate` does not work without hacks in the same process and its also an expensive task and should not delay the import and obscure errors and swallow output if done as part of the import.

mhsdesign committed:
TASK: Assert that no events exist on the live content stream ... but later

dlubitz committed:
Merge remote-tracking branch 'origin/9.0' into feature/4448-site-import

dlubitz committed:
TASK: Post-merge fix of tests

bwaidelich commented on Neos.ContentRepository.LegacyNodeMigration/Tests/Behavior/Bootstrap/FeatureContext.php:
```diff
@@ -116,20 +86,19 @@ public function iHaveTheFollowingNodeDataRows(TableNode $nodeDataRows): void
+     * @When I run the event migration for workspace :workspace
      */
-    public function iRunTheEventMigration(string $contentStream = null, RootNodeTypeMapping $rootNodeTypeMapping = null): void
+    public function iRunTheEventMigration(string $workspace = null, RootNodeTypeMapping $rootNodeTypeMapping = null): void
```
The `$workspace` parameter is not used

mficzel commented:
We could use a check for doctrine:migrate and cr:setup as it is important that those were applied but not run them automagically.

mhsdesign commented:
Jup @mficzel that were exactly my thoughts as well. Implemented that in:

https://github.com/neos/neos-development-collection/pull/5307/commits/adf7b38f26966ae5440578102e2700d0d3b7be33

I also discussed this with denny and we agree on that the schema creation should be dedicated.

mhsdesign committed:
TASK: Remove obsolete `I run the event migration for workspace :workspace`

bwaidelich commented:
> We could use a check for doctrine:migrate and cr:setup as it is important that those were applied but not run them automagically.

I'd be fine with that. As long as there is no cryptic exception when forgetting to execute a number of commands in a very specific order ;)

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Classes/Command/SiteCommandController.php:
```diff
@@ -18,50 +18,39 @@
      * @throws \Exception
      */
-    public function migrateLegacyDataCommand(bool $verbose = false, string $config = null): void
+    public function migrateLegacyDataCommand(string $contentRepository = 'default', bool $verbose = false, string $config = null): void
```
i wonder is it of any help to do this in one step and have no in between results?

Currently if anything fails in the `importFromPath` we dont even unlink the `$temporaryFilePath` and it would be frustrating i would have nothing agains using the new `exportLegacyData` only like:

```
flow site:exportLegacyDataCommand --path /temp/but/the/user/sees/it
flow site:importAll --path /temp/but/the/user/sees/it
```

Otherwise we would need to check all preconditions already in `importAll` like is doctrine migrated etc ...
I think it would be good to enforce that transparency and make things easier for us (less duplicated code ..., less long running processes where steps could fail ANYWHERE) no?

bwaidelich replied:
As discussed: I think it's a good idea to separate the two, i.e. instead of migrating directly to export legacy sites into the new import format and then use the same entry point for both ways to _import_ a new site

mhsdesign committed:
TASK: Adjust documentation and todos in legacy migration

mhsdesign committed:
TASK: Remove obsolete `ProjectionReplayProcessor`

mhsdesign reviewed (approved):
Giving this +.5 already ;) will need to test that again with the Neos ui e2e tests (as they use the export as well) but for now it looks really good. We also have to adjust the `flow setup` to guide to these new commands :)

As written here https://github.com/neos/neos-development-collection/pull/5307#discussion_r1834934970 im really questioning if `migrateLegacyData` should be a one step. We alreay require that people migrate to the `Neos.Neos:Site` homepage nodetype and their dimension config either way.

One more - good imo - step in between would just allow more flexibility and "come back later" to work on the migration.

mhsdesign committed:
BUGFIX: Import `Duplicate entry 'onedimension.localhost' for key 'flow_identity_neos_neos_domain_model_domain`

Instead, we will now gracefully handle this case:

Domain "onedimension.localhost" already exists. Adding it to site "Neos Test Site".

mhsdesign committed:
TASK: Migration, make sure that NodeType does not exist error is thrown first

mhsdesign commented:
> I'd be fine with that. As long as there is no cryptic exception when forgetting to execute a number of commands in a very specific order ;)

this would be it but could further be improved ... but idk ... :D Its just the flow command line exception handling printer :)

<img width="768" alt="image" src="https://github.com/user-attachments/assets/2cf9fb0e-ff95-4db8-83e1-5e62310b5bef">

----

also im just testing this and it works so far, i guess its okay that the export overrides the directory specified but could also be dangerous?
maybe:

```php
$directoryContainsFiles = (new \FilesystemIterator($path))->valid();
if ($directoryContainsFiles) {
    $this->outputLine(sprintf('Path %s is not empty.', $path));
    $this->quit(1);
}
```

mhsdesign committed:
TASK: Enforce file creation first (to throw possibly errors)

mhsdesign committed:
TASK: Dont use `findOneByHost` because it returns not the exact result

In our case we want to skip the creation of a domain if it exists based on the host name (see c1d64ba3f4ccf01f60fd4f3f20f93a1642ba3ff2) but now we always skip sub domains, if the main domain is created first.

mhsdesign committed:
BUGFIX: Create site with correct online state (inverse condition)

mhsdesign committed:
TASK: Simplify `SiteShape`

mhsdesign committed:
TASK: Remove `site:migrateLegacyData` (previously `cr:migrateLegacyData`)

replaced with

```
./flow site:exportLegacyDataCommand --path ./migratedContent
./flow site:importAll --path ./migratedContent
```

mhsdesign committed:
BUGFIX: Export correctly export `primary` domain

previously `primary` was always true if there is only one domain

mhsdesign committed:
BUGFIX: `site:importAll` import `copyrightNotice` of assets

previously we didnt use the value that was in the export format.

mhsdesign committed:
TASK: Trivial cosmetic changes

mhsdesign committed:
TASK: Make pruning a noop if the workspace does not exist

mhsdesign committed:
TASK: Adjust documentation to reference correct command

mhsdesign committed:
TASK: Adjust test to 49234ffb1bd3a5f3391120f8eb0be72c6d831952

mhsdesign reviewed (approved):
Okay i spend a lotta time testing this extensively now and fixing bugs and edgecases.

I tested doing a `exportLegacyData` with my local Neos 8 demo + domain and another site i use for Neos UI tests.
Then i imported that one again into Neos 9 testet the result via ui.
And to make it a full circle i exported that result again via `exportAll` and diffed the folders as we expect the same result - which it now finally is.

Following fixed bugs:
- imported sites are not active
- primary domain export was not correct
- having the same domain already was not a noop but crashed due to constraints
- fail early if `--path` is not a directory
- respect `copyrightNotice` of serialised assets during import

Now im at the state to say we can merge :D

bwaidelich commented:
Thanks a lot for putting so much effort into this one, everybody!
Let's get this beast merged at some point. We can always tweak it later

mhsdesign commented:
As per martin from slack

> Fein

:D lets merge!

kitsunet commented:
UI tests are now updated.

mhsdesign commented:
Lol i just found an early draft in our git history https://github.com/neos/neos-development-collection/commit/eaf246cd313b4650fab749913ea9ac1de858acbd the export thing was now reworked also three times 😅 :D
